### ISSUE TITLE
Collect, document, and answer FAQ

### ISSUE THREAD

**Issue by SergioBenitez** [SergioBenitez/Rocket#1585]
There are questions that are frequently asked that are not directly related to Rocket but impact Rocketeers nevertheless. In pursuit of making Rocket as friendly and easy to use, we should collect these questions in an official capacity and answer them once and for all. This issue tracks such questions.

  * [ ] How do I listen on IPv4 and IPv6?
  * [x] How do I upload a file / what is this multipart metadata in my stream?
  * [x] Can I, and if so how, do I handle websockets?
  * [x] How do I respond with different kinds of responses, with different status codes?
  * [x] Should I use global state with something like `lazy_static!`?
  * [x] How do I access managed state outside of a Rocket-related context?
  * [ ] How do I mount a generic route?
  * [x] How do I make Rocket restart automatically when I change source code?
  * [x] How do I get an `&Request` in a handler?
  * [ ] How do I get a specific header in a handler?
  * [x] When will version `x + 1` be released? Why does it take so long?
  * [x] Is example `foo` broken? It doesn't work for me.
  * [ ] Why so many macros? I heard they lead to bad errors and make my code harder to understand.
  * [x] Is Rocket slow? Is Rocket fast?
  * [x] Can I use Rocket in production? Should I? It's only 0.x!
 
Please feel free to make suggestions!

**Comment by ThouCheese** [comment]
I would suggest adding "why doesn't writing `rocket = "0.5.0-dev"` in Cargo.toml work

**Comment by jebrosen** [comment]
* How do I get the value of a request header (e.g. `Host`, `User-Agent`) in a route?

**Comment by Flying-Toast** [comment]
- Why is {x} in Rocket slower than {other framework}

**Comment by SergioBenitez** [comment]
Hiding comments as "resolved" as I bring them into the top-level post or they get addressed elsewhere in docs.

**Comment by Darthfett** [comment]
- How do I configure Rocket to handle SSL traffic (HTTPS)?

**Comment by Xunjin** [comment]
How can I structure a more "long term" project? Focusing on maintenance and segment concerns (maybe some example in "known" software architectures). 

PS: That would probably be a book/set of articles itself.

**Comment by jhwgh1968** [comment]
* How do I start and manage rocket from non-async Rust code?

This is the question I had while upgrading a rather complex project from 0.4.

In this project, a lot of other stuff was going on, and the HTTP server was only a secondary function enabled when requested. Converting the rest of my code to async Rust was simply too much work.

Instead, I wrote a `start_server` function to replicate the conditional config and blocking startup in pre-async Rocket. But I feel pretty certain that ugly hack is the wrong approach.

Not only does it not use the new `#[launch]` macro anywhere, but I had to use `unsafe` code to actually return errors properly. (I wanted to get the `std::io::Error` from a `rocket::Error` of `Kind::Io` by value, but the `ErrorKind` that contains it is returned by reference.)

I imagine a Migration Guide will need to be written at some point, and I would offer to provide more details elsewhere if it would help with that effort. This is just to register that I think a question providing a hint/official pointer should be in the FAQ.
